File: python/scripts/zmq_loopback_command.py
import argparse
import time

import zmq
from state_representation import CartesianState, JointState

from network_interfaces.zmq import network


def main(state_uri, command_uri):
    context = zmq.Context(1)
    subscriber = network.configure_subscriber(context, command_uri, False)
    publisher = network.configure_publisher(context, state_uri, False)

    state = network.StateMessage()
    state.joint_state = JointState().Random("loopback", 7)
    state.ee_state = CartesianState().Random("loopback_ee")

    while True:
        network.send_state(state, publisher)
        command = network.receive_command(subscriber)
        if command:
            print(command)
        time.sleep(0.1)

    exit(0)


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("state_uri", type=str, help="URI of the state socket")
    parser.add_argument("command_uri", type=str, help="URI of the command socket")
    args = parser.parse_args()

    main(args.state_uri, args.command_uri)

File: cpp/scripts/zmq_loopback_command.cpp
#include <unistd.h>
#include <state_representation/space/cartesian/CartesianState.hpp>
#include <state_representation/space/joint/JointState.hpp>

#include "network_interfaces/zmq/network.h"

int main(int argc, char** argv) {
  std::string state_uri, command_uri;
  if (argc != 3) {
    std::cerr << "Provide two arguments: state_uri command_uri" << std::endl;
    return 1;
  } else {
    state_uri = argv[1];
    command_uri = argv[2];
  }
  ::zmq::context_t context(1);
  ::zmq::socket_t subscriber, publisher;
  network_interfaces::zmq::configure_subscriber(context, subscriber, command_uri, false);
  network_interfaces::zmq::configure_publisher(context, publisher, state_uri, false);

  network_interfaces::zmq::StateMessage state;
  state.joint_state = state_representation::JointState::Random("loopback", 7);
  state.ee_state = state_representation::CartesianState::Random("loopback_ee");
  network_interfaces::zmq::CommandMessage command;

  while (true) {
    network_interfaces::zmq::send(state, publisher);
    if (network_interfaces::zmq::receive(command, subscriber)) {
      std::cout << command << std::endl;
    }
    usleep(100000);
  }

  return 0;
}
